File: SuperLauncher/network/api_client.py
import requests
import threading
from typing import Callable, Optional

class ApiClient:
    """Async HTTP client with callbacks"""

    def get(self, url: str, callback: Callable = None, error_callback: Callable = None,
            headers: dict = None, timeout: int = 15, **params):
        def task():
            try:
                h = headers or {"User-Agent": "SuperLauncher/2.0"}
                if params:
                    resp = requests.get(url, params=params, headers=h, timeout=timeout)
                else:
                    resp = requests.get(url, headers=h, timeout=timeout)
                resp.raise_for_status()
                if callback:
                    callback(resp.json())
            except Exception as e:
                if error_callback:
                    error_callback(str(e))
        threading.Thread(target=task, daemon=True).start()

    def get_raw(self, url: str, callback: Callable = None, error_callback: Callable = None,
                headers: dict = None, timeout: int = 15):
        def task():
            try:
                h = headers or {"User-Agent": "SuperLauncher/2.0"}
                resp = requests.get(url, headers=h, timeout=timeout)
                resp.raise_for_status()
                if callback:
                    callback(resp.content)
            except Exception as e:
                if error_callback:
                    error_callback(str(e))
        threading.Thread(target=task, daemon=True).start()

    def post(self, url: str, data: dict, callback: Callable = None, error_callback: Callable = None,
             headers: dict = None, timeout: int = 15):
        def task():
            try:
                h = headers or {"User-Agent": "SuperLauncher/2.0", "Content-Type": "application/json"}
                resp = requests.post(url, json=data, headers=h, timeout=timeout)
                resp.raise_for_status()
                if callback:
                    callback(resp.json())
            except Exception as e:
                if error_callback:
                    error_callback(str(e))
        threading.Thread(target=task, daemon=True).start()

    @staticmethod
    def download_file(url: str, save_path: str, progress_callback: Callable = None):
        resp = requests.get(url, stream=True, timeout=120)
        resp.raise_for_status()
        total = int(resp.headers.get("content-length", 0))
        written = 0
        with open(save_path, "wb") as f:
            for chunk in resp.iter_content(8192):
                if chunk:
                    f.write(chunk)
                    written += len(chunk)
                    if progress_callback and total > 0:
                        progress_callback(int(written * 100 / total))
        if progress_callback:
            progress_callback(100)

File: SuperLauncherCpp/src/network/ApiClient.cpp
#include "ApiClient.h"
#include <QNetworkRequest>
#include <QFile>
#include <QUrl>

ApiClient::ApiClient(QObject *parent)
    : QObject(parent)
    , m_manager(new QNetworkAccessManager(this))
{
}

void ApiClient::setDefaultHeader(const QString &key, const QString &value)
{
    m_defaultHeaders[key] = value;
}

void ApiClient::get(const QString &url, Callback cb,
                    const QMap<QString, QString> &headers)
{
    QUrl theUrl(url);
    QNetworkRequest req(theUrl);
    req.setRawHeader("User-Agent", "SuperLauncher/2.0");
    req.setTransferTimeout(15000);

    for (auto it = m_defaultHeaders.begin(); it != m_defaultHeaders.end(); ++it)
        req.setRawHeader(it.key().toUtf8(), it.value().toUtf8());
    for (auto it = headers.begin(); it != headers.end(); ++it)
        req.setRawHeader(it.key().toUtf8(), it.value().toUtf8());

    auto *reply = m_manager->get(req);
    connect(reply, &QNetworkReply::finished, this, [reply, cb]() {
        if (reply->error() == QNetworkReply::NoError) {
            QJsonDocument doc = QJsonDocument::fromJson(reply->readAll());
            cb(true, doc);
        } else {
            cb(false, QJsonDocument());
        }
        reply->deleteLater();
    });
}

void ApiClient::get(const QString &url, Callback cb,
                    ProgressCallback prog,
                    const QMap<QString, QString> &headers)
{
    QUrl theUrl(url);
    QNetworkRequest req(theUrl);
    req.setRawHeader("User-Agent", "SuperLauncher/2.0");
    for (auto it = m_defaultHeaders.begin(); it != m_defaultHeaders.end(); ++it)
        req.setRawHeader(it.key().toUtf8(), it.value().toUtf8());
    for (auto it = headers.begin(); it != headers.end(); ++it)
        req.setRawHeader(it.key().toUtf8(), it.value().toUtf8());

    auto *reply = m_manager->get(req);
    if (prog)
        connect(reply, &QNetworkReply::downloadProgress, this, prog);
    connect(reply, &QNetworkReply::finished, this, [reply, cb]() {
        if (reply->error() == QNetworkReply::NoError) {
            QJsonDocument doc = QJsonDocument::fromJson(reply->readAll());
            cb(true, doc);
        } else {
            cb(false, QJsonDocument());
        }
        reply->deleteLater();
    });
}

void ApiClient::post(const QString &url, const QJsonObject &body, Callback cb,
                     const QMap<QString, QString> &headers)
{
    QUrl theUrl(url);
    QNetworkRequest req(theUrl);
    req.setRawHeader("User-Agent", "SuperLauncher/2.0");
    req.setHeader(QNetworkRequest::ContentTypeHeader, "application/json");
    for (auto it = m_defaultHeaders.begin(); it != m_defaultHeaders.end(); ++it)
        req.setRawHeader(it.key().toUtf8(), it.value().toUtf8());
    for (auto it = headers.begin(); it != headers.end(); ++it)
        req.setRawHeader(it.key().toUtf8(), it.value().toUtf8());

    QByteArray payload = QJsonDocument(body).toJson(QJsonDocument::Compact);
    auto *reply = m_manager->post(req, payload);
    connect(reply, &QNetworkReply::finished, this, [reply, cb]() {
        if (reply->error() == QNetworkReply::NoError) {
            QJsonDocument doc = QJsonDocument::fromJson(reply->readAll());
            cb(true, doc);
        } else {
            cb(false, QJsonDocument());
        }
        reply->deleteLater();
    });
}

void ApiClient::getRaw(const QString &url,
                        std::function<void(bool, const QByteArray &)> cb,
                        const QMap<QString, QString> &headers)
{
    QUrl theUrl(url);
    QNetworkRequest req(theUrl);
    req.setRawHeader("User-Agent", "SuperLauncher/2.0");
    for (auto it = m_defaultHeaders.begin(); it != m_defaultHeaders.end(); ++it)
        req.setRawHeader(it.key().toUtf8(), it.value().toUtf8());
    for (auto it = headers.begin(); it != headers.end(); ++it)
        req.setRawHeader(it.key().toUtf8(), it.value().toUtf8());

    auto *reply = m_manager->get(req);
    connect(reply, &QNetworkReply::finished, this, [reply, cb]() {
        if (reply->error() == QNetworkReply::NoError) {
            cb(true, reply->readAll());
        } else {
            cb(false, QByteArray());
        }
        reply->deleteLater();
    });
}

void ApiClient::uploadRaw(const QString &url, const QByteArray &data,
                           const QString &contentType, Callback cb)
{
    QUrl theUrl(url);
    QNetworkRequest req(theUrl);
    req.setRawHeader("User-Agent", "SuperLauncher/2.0");
    req.setRawHeader("Content-Type", contentType.toUtf8());

    auto *reply = m_manager->post(req, data);
    connect(reply, &QNetworkReply::finished, this, [reply, cb]() {
        if (reply->error() == QNetworkReply::NoError) {
            QJsonDocument doc = QJsonDocument::fromJson(reply->readAll());
            cb(true, doc);
        } else {
            cb(false, QJsonDocument());
        }
        reply->deleteLater();
    });
}

void ApiClient::downloadFile(const QString &url, const QString &savePath,
                             Callback cb, ProgressCallback prog)
{
    QUrl theUrl(url);
    QNetworkRequest req(theUrl);
    req.setRawHeader("User-Agent", "SuperLauncher/2.0");

    auto *reply = m_manager->get(req);
    if (prog)
        connect(reply, &QNetworkReply::downloadProgress, this, prog);

    connect(reply, &QNetworkReply::finished, this, [reply, savePath, cb]() {
        if (reply->error() == QNetworkReply::NoError) {
            QFile f(savePath);
            if (f.open(QIODevice::WriteOnly)) {
                f.write(reply->readAll());
                f.close();
                cb(true, QJsonDocument());
            } else {
                cb(false, QJsonDocument());
            }
        } else {
            cb(false, QJsonDocument());
        }
        reply->deleteLater();
    });
}
